astropy/cosmology/funcs.py

```diff
@@ -90,56 +125,130 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
+            warnings.warn(f"Option 'bracket' is ignored by method {method}.")
+            bracket = None
+    else:
+        opt['xtol'] = ztol
+
+    # fval falling inside the interval of bracketing function values does not guarantee
+    # it has a unique solution, but for Standard Cosmological quantities normally should
+    # (being monotonic or having a single extremum).
+    # In these cases keep solver from returning solutions outside of bracket.
+    fval_zmin, fval_zmax = func((zmin, zmax))
+    nobracket = False
     if np.sign(fval - fval_zmin) != np.sign(fval_zmax - fval):
-        warnings.warn("""\
-fval is not bracketed by func(zmin) and func(zmax). This means either
-there is no solution, or that there is more than one solution between
-zmin and zmax satisfying fval = func(z).""")
+        if bracket is None:
+            nobracket = True
+        else:
+            fval_brac = func(bracket)
+            if np.sign(fval - fval_brac[0]) != np.sign(fval_brac[-1] - fval):
+                nobracket = True
+            else:
+                zmin, zmax = bracket[0], bracket[-1]
+                fval_zmin, fval_zmax = fval_brac[[0, -1]]
+    if nobracket:
+        warnings.warn(f"fval is not bracketed by func(zmin)={fval_zmin} and func(zmax)="
+                      f"{fval_zmax}. This means either there is no solution, or that there is "
+                      "more than one solution between zmin and zmax satisfying fval = func(z).",
+                      AstropyUserWarning)
 
     if isinstance(fval_zmin, Quantity):
         val = fval.to_value(fval_zmin.unit)
-        f = lambda z: abs(func(z).value - val)
     else:
-        f = lambda z: abs(func(z) - fval)
-
-    zbest, resval, ierr, ncall = fminbound(f, zmin, zmax, maxfun=maxfun,
-                                           full_output=1, xtol=ztol)
-
-    if ierr != 0:
-        warnings.warn(f'Maximum number of function calls ({ncall}) reached')
-
-    if np.allclose(zbest, zmax):
-        raise CosmologyError("Best guess z is very close the upper z limit.\n"
+        val = fval
+
+    # 'Brent' and 'Golden' ignore `bounds`, force solution inside zlim
+    def f(z):
+        if z > zmax:
+            return 1.e300 * (1.0 + z - zmax)
+        elif z < zmin:
+            return 1.e300 * (1.0 + zmin - z)
+        elif isinstance(fval_zmin, Quantity):
+            return abs(func(z).value - val)
+        else:
+            return abs(func(z) - val)
+
+    res = minimize_scalar(f, method=method, bounds=(zmin, zmax), bracket=bracket, options=opt)
+
+    # Scipy docs state that `OptimizeResult` always has 'status' and 'message' attributes,
+    # but only `_minimize_scalar_bounded()` seems to have really implemented them.
+    if not res.success:
+        warnings.warn(f"Solver returned {res.get('status')}: {res.get('message', 'Unsuccessful')}\n"
+                      f"Precision {res.fun} reached after {res.nfev} function calls.")
+
+    if verbose:
+        print(res)
```

Comment:
In the tests, this line can be covered by

```
import sys
original_stdout = sys.stdout # Save a reference to the original standard output
with open('filename.txt', 'w') as f:
    sys.stdout = f # Change the standard output to the file we created.
    print('This message will be written to a file.')
    sys.stdout = original_stdout # Reset the standard output to its original value
```
(https://www.stackabuse.com/writing-to-a-file-with-pythons-print-function/)

Of course, we'd want a temporary file instead.